sdk/trace/src/main/java/io/opentelemetry/sdk/trace/AttributesMap.java

```diff
@@ -52,32 +44,22 @@ int getTotalAddedValues() {
   @SuppressWarnings("unchecked")
   @Override
   public <T> T get(AttributeKey<T> key) {
-    return (T) data.get(key);
-  }
-
-  @Override
-  public int size() {
-    return data.size();
-  }
-
-  @Override
-  public boolean isEmpty() {
-    return data.isEmpty();
+    return (T) super.get(key);
   }
 
   @SuppressWarnings({"rawtypes", "unchecked"})
   @Override
-  public void forEach(BiConsumer<AttributeKey<?>, Object> consumer) {
-    for (Map.Entry<AttributeKey<?>, Object> entry : data.entrySet()) {
+  public void forEach(BiConsumer<? super AttributeKey<?>, ? super Object> consumer) {
+    for (Map.Entry<AttributeKey<?>, Object> entry : entrySet()) {
       AttributeKey key = entry.getKey();
       Object value = entry.getValue();
-      consumer.accept(key, value);
+      consumer.accept((AttributeKey<?>) key, value);
     }
   }
 
   @Override
   public Map<AttributeKey<?>, Object> asMap() {
-    return Collections.unmodifiableMap(data);
+    return Collections.unmodifiableMap(this);
```

Comment:
IntelliJ flags this as "redundant usage of unmodifiableMap wrapper" but I can't find anything wrong with it, without the wrapper the user would be able to mutate this map

Reply:
`AttributesMap` implements `Attributes` which is marked `@Immutable`. If that were true, the mapper would indeed be redundant.

Reply:
Ah thanks for explaining! Yeah this is an implementation detail to support `Attributes.builder().putAll(...)` and not something that leaks into the API, so I guess I just need to add a comment.

Reply:
Looking at https://errorprone.info/bugpattern/Immutable, `@Immutable` on an interface is supposed to mean that all classes implementing it are `@Immutable`. It does not only mean that mutation through the interface is impossible. So I think we need to remove that annotation from `Attributes`.

Reply:
Well, we also say explicitly

```
Implementations of this interface *must* be immutable and have well-defined value-based
 * equals/hashCode implementations. If an implementation does not strictly conform to these
 * requirements, behavior of the OpenTelemetry APIs and default SDK cannot be guaranteed.
```

but ignore our own advice because it's an implementation detail that worked ok for us. If it's too weird, we could remove `implements Attributes` and reimplement `putAll` instead, I guess it's not all that different.

Reply:
ok, Agrawal. I'll let this one through, but I'm watching you like a hawk for abuses of this antipattern.  👁️ 👁️  🐦